modules/assim.batch/R/pda.emulator.ms.R

```diff
@@ -0,0 +1,518 @@
+      
+      if(!is.null(sf)){
+        sf.samp.list[[c]] <- sfm
+      }
+      
+      resume.list[[c]] <- mcmc.out[[c]]$chain.res
+    }
+    
+    # Separate each PFT's parameter samples (and bias term) to their own list
+    mcmc.param.list <- list()
+    ind <- 0
+    for (i in seq_along(n.param.orig)) {
+      mcmc.param.list[[i]] <- lapply(mcmc.samp.list, function(x) x[, (ind + 1):(ind + n.param.orig[i]), drop = FALSE])
+      ind <- ind + n.param.orig[i]
+    }
+    
+    # Collect non-model parameters in their own list
+    if(length(mcmc.param.list) > length(tmp.settings$pfts)) { 
+      # means bias parameter was at least one bias param in the emulator
+      # it will be the last list in mcmc.param.list
+      # there will always be at least one tau for bias
+      for(c in seq_len(tmp.settings$assim.batch$chain)){
+        mcmc.param.list[[length(mcmc.param.list)]][[c]] <- cbind( mcmc.param.list[[length(mcmc.param.list)]][[c]],
+                                                                  mcmc.out[[c]]$mcmc.par)
+      }
+      
+    } else if (ncol(mcmc.out[[1]]$mcmc.par) != 0){
+      # means no bias param but there are still other params, e.g. Gaussian
+      mcmc.param.list[[length(mcmc.param.list)+1]] <- list()
+      for(c in seq_len(tmp.settings$assim.batch$chain)){
+        mcmc.param.list[[length(mcmc.param.list)]][[c]] <- mcmc.out[[c]]$mcmc.par
+      }
+    }
+    
+    tmp.settings <- pda.postprocess(tmp.settings, con, mcmc.param.list, pname, prior.list, prior.ind.orig, sffx = "_global")
+    
+  } # global - if end
+  
+  ## -------------------------------------- Hierarchical MCMC ------------------------------------------ 
+  if(hierarchical){ # hierarchical - if begin
+    
+    ## Get an ensemble id for hierarchical calibration
+    tmp.settings$assim.batch$ensemble.id <- pda.create.ensemble(tmp.settings, con, workflow.id)
+    
+    ## Transform values from non-normal distributions to standard Normal
+    ## it won't do anything if all priors are already normal
+    norm_transform <- norm_transform_priors(prior.list, prior.fn.all, prior.ind.all, SS.stack, init.list, jmp.list)
+    if(!norm_transform$normF){ # means SS values are transformed
+      ## get new SS.stack with transformed values
+      SS.stack <- norm_transform$normSS
+      
+      ## re-fit GP on new param space
+      for(i in seq_along(SS.stack)){
+        GPmodel <- lapply(SS.stack[[i]], function(x) mlegp::mlegp(X = x[, -ncol(x), drop = FALSE], Z = x[, ncol(x), drop = FALSE], nugget = 0, nugget.known = 1, verbose = 0))
+        gp.stack[[i]] <- GPmodel
+      }
+      
+      ## re-define rng
+      rng <- norm_transform$rng
+      
+      ## get new init.list and jmp.list
+      init.list <- norm_transform$init
+      jmp.list  <- norm_transform$jmp
+
+      
+    }
+    
+
+    ########### hierarchical MCMC function with Gibbs ##############
+    
+    
+    hier.mcmc <- function(settings, gp.stack, nstack, nmcmc, rng, 
```

Comment:
the only reason this is a function was that so that I can use `parLapply` on it later and I kinda found it convenient to have it in the same script while developing but now I can move it to its own document